test/type_construction_test.rb

```diff
@@ -6482,4 +6482,94 @@ def test_typing_record_nilable_attribute
       end
     end
   end
+
+  def test_type_case_case_when_assignment
+    with_checker do |checker|
+      source = parse_ruby(<<EOF)
+# @type var x: String | Integer
+x = (_ = nil)
+
+case x
+when String
+  a = "String"
+when Integer
+  a = "Integer"
+end
+EOF
+
+      with_standard_construction(checker, source) do |construction, typing|
+        _, _, context = construction.synthesize(source.node)
+
+        assert_no_error typing
+        assert_equal parse_type("::String"), context.lvar_env[:a]
+      end
+    end
+  end
+
+  def test_type_case_case_selector
```

Comment:
Variable type narrowing didn't work well if the condition of case-when is an assignment.